ipn/ipnlocal/profiles.go

```diff
@@ -0,0 +1,467 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package ipnlocal
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"runtime"
+	"sync"
+	"time"
+
+	"golang.org/x/exp/slices"
+	"tailscale.com/ipn"
+	"tailscale.com/types/logger"
+	"tailscale.com/util/strs"
+)
+
+// ProfileManager is a wrapper around a StateStore that manages
+// multiple profiles and the current profile.
+type ProfileManager struct {
+	store ipn.StateStore
+	logf  logger.Logf
+
+	// Lock order: LocalBackend.mu, then pm.mu.
+	mu             sync.Mutex // guards following
+	currentUserID  string     // only used on Windows
+	knownProfiles  map[ipn.ProfileID]*ipn.LoginProfile
+	currentProfile *ipn.LoginProfile
+	prefs          ipn.PrefsView
+
+	// isNewProfile is a sentinel value that indicates that the
+	// current profile is new and has not been saved to disk yet.
+	// It is reset to false after a call to SetPrefs with a filled
+	// in LoginName.
+	isNewProfile bool
+}
+
+// CurrentUser returns the current user ID. It is only non-empty on
+// Windows where we have a multi-user system.
+func (pm *ProfileManager) CurrentUser() string {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	return pm.currentUserID
+}
+
+// SetCurrentUser sets the current user ID. The uid is only non-empty
+// on Windows where we have a multi-user system.
+func (pm *ProfileManager) SetCurrentUser(uid string) error {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	if pm.currentUserID == uid {
+		return nil
+	}
+	cpk := ipn.CurrentProfileKey(uid)
+	if b, err := pm.store.ReadState(cpk); err == nil {
+		pk := ipn.StateKey(string(b))
+		prefs, err := pm.loadSavedPrefs(pk)
+		if err != nil {
+			return err
+		}
+		pm.currentProfile = pm.findProfileByKeyLocked(pk)
+		pm.prefs = prefs
+		pm.isNewProfile = false
+	} else if err == ipn.ErrStateNotExist {
+		pm.prefs = emptyPrefs
+		pm.isNewProfile = true
+	} else {
+		return err
+	}
+	pm.currentUserID = uid
+	return nil
+}
+
+func (pm *ProfileManager) findProfileByNameLocked(name string) *ipn.LoginProfile {
+	for _, p := range pm.knownProfiles {
+		if p.Name == name {
+			return p
+		}
+	}
+	return nil
+}
+
+func (pm *ProfileManager) findProfileByKeyLocked(key ipn.StateKey) *ipn.LoginProfile {
+	for _, p := range pm.knownProfiles {
+		if p.Key == key {
+			return p
+		}
+	}
+	return nil
+}
+
+func (pm *ProfileManager) setUnattendedModeAsConfigured() error {
+	if pm.currentUserID == "" {
+		return nil
+	}
+
+	if pm.prefs.ForceDaemon() {
+		return pm.store.WriteState(ipn.ServerModeStartKey, []byte(pm.currentProfile.Key))
+	} else {
+		return pm.store.WriteState(ipn.ServerModeStartKey, nil)
+	}
+}
+
+// Reset unloads the current profile, if any.
+func (pm *ProfileManager) Reset() {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	pm.prefs = emptyPrefs
+	pm.currentUserID = ""
+	pm.currentProfile = nil
+}
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+func randID() string {
```

Comment:
Maybe call this `randProfileID` and have it return a `ProfileID` type directly?

Reply:
(can also inline it in `newUnusedKey`, since it's only used once)

Reply:
Done